common/membership/hashring.go

```diff
@@ -99,7 +99,7 @@ func newHashring(
 		service:      service,
 		peerProvider: provider,
 		shutdownCh:   make(chan struct{}),
-		refreshChan:  make(chan *ChangedEvent),
+		refreshChan:  make(chan struct{}, 1),
```

Comment:
I know the log you're referring to the in the description, but I'm slightly nervous changing this: I'd want to ensure that it doesn't cause some upstream process to block way more suddenly. 

Have we tested it in practice? I'm guessing the result would be to drop just a lot more events, or block a lot more? 

Given that we don't really care about the changes, only the whole state, would a debounce (on an event, waiting for 100 MS and then only processing the last event and dropping all the rest for example) more sense?

Reply:
oh, I think there's already a debounce (line 255)

I'm not sure I understand why we should try to make this blocking then? Not opposed to the change, but apart from the log noise, I'm not sure if it presents a problem?

Reply:
Let me explain this since it is really tricky part. What we currently have in hashring.go is unbuffered channel. Writes from peerProvider (ringpop/uns) are already happening in a non-blocking way (via select). It means write will **only succeed** if receiver is reading the channel right now. If it is doing something else (not ready yet, or it is in `refresh()` which ofc takes some time) - write will fail. That's why we see huge % of writes are failing - another membership change happens during `refresh()` or update or notifying subscribers.

I made this playground to illustrate the issue: https://go.dev/play/p/2Os6fkarH8W
Having a channel of size 1 guarantees we **never miss the fact of being notified** (notice, PeerProvider now calls an update function).

What happened before (with non-buffered chan):
1. PeerProvider writes ChangedEvent
2. We received ChangedEvent and called refresh() etc.
3. While we were in refresh() PeerProvider got another event and tried  to write it to the channel. Since we're not reading it right now (we are in refresh()), it won't be able to notify us, and will log "channel is full"
4. We finished refresh(), but we don't have anything in the channel, so we will wait until the next refresh() or defaultRefreshInterval=10s to capture the update.

What happens now (with buffer=1):
1. PeerProvider calls handler
2. handler adds an event to channel to call "refresh()"
3. We read refreshChan; channel capacity is now empty; we are calling called refresh() etc. 
4. While we were in refresh() PeerProvider got another event and calls handler again. Handler will add notification to the refreshChan.
5. refresh() and others are finished, we back to reading refreshChan. There is a message to read; which means, we didn't loose the fact we've been notified.

Reply:
>why we should try to make this blocking then?
What do you mean by that?

Reply:
my comment was wrong, I follow your point after after discussing: That we risk loosing the last event if they're sequenced one after the other and this keeps diverting to the default path. 🙏  for the explanation